platforms/core-runtime/service-registry-impl/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -180,31 +181,36 @@ public String toString() {
         return getDisplayName();
     }
 
-    private void findProviderMethods(ServiceRegistrationProvider target) {
+    private void findProviderMethods(ServiceRegistrationProvider target, ServiceAccessToken token) {
         Class<? extends ServiceRegistrationProvider> type = target.getClass();
         RelevantMethods methods = RelevantMethods.getMethods(type);
         for (ServiceMethod method : methods.decorators) {
             if (parentServices == null) {
                 throw new ServiceLookupException(String.format("Cannot use decorator method %s.%s() when no parent registry is provided.", type.getSimpleName(), method.getName()));
             }
-            ownServices.add(new FactoryMethodService(this, target, method));
+            ownServices.add(new FactoryMethodService(this, determineAccessScope(method, token), token, target, method));
         }
         for (ServiceMethod method : methods.factories) {
-            ownServices.add(new FactoryMethodService(this, target, method));
+            ownServices.add(new FactoryMethodService(this, determineAccessScope(method, token), token, target, method));
         }
         for (ServiceMethod method : methods.configurers) {
-            applyConfigureMethod(method, target);
+            applyConfigureMethod(token, method, target);
         }
     }
 
-    private void applyConfigureMethod(ServiceMethod method, Object target) {
+    private static ServiceAccessScope determineAccessScope(ServiceMethod method, ServiceAccessToken token) {
+        PrivateService privateService = method.getMethod().getAnnotation(PrivateService.class);
+        return privateService != null ? ServiceAccess.getPrivateScope(token) : ServiceAccess.getPublicScope();
+    }
```

Comment:
ℹ️ If the `@PrivateService` annotation is present on the provider method, then it gets a private scope that corresponds to the services registered via the same service provider. Otherwise, the service gets a public scope that is visible from everywhere, matching the previous behavior.